client/service/src/client/client.rs

```diff
@@ -974,12 +995,15 @@ where
 		};
 
 		if let Some(storage_changes) = storage_changes {
-			// TODO [ToDr] How to handle re-orgs? Should we re-emit all storage changes?
-			self.storage_notifications.trigger(
-				&notification.hash,
-				storage_changes.0.into_iter(),
-				storage_changes.1.into_iter().map(|(sk, v)| (sk, v.into_iter())),
-			);
+			self.all_storage_changes.trigger(&notification.hash, storage_changes.0.clone());
+
+			if notification.is_new_best {
+				self.storage_changes_for_keys.lock().trigger(
+					notification.header.parent_hash(),
+					&notification.hash,
+					storage_changes.0,
+				);
+			}
```

Comment:
I just realized that all of this is not going to work for parachains :see_no_evil: The problem being that for Parachains we don't import them as best block directly. We first import the block, aka execute it. Then we would have `storage_changes` being set and then we import the block again when it is the best block (aka got enacted by the relay chain). But the second time there are no more storage changes, so we would never update the listeners.

Reply:
That's good to know! I'll be sure to add a test for this scenario.

Hmm.... so I suppose we could fix this by ignoring what the `is_new_best` says here and just calculate it on our own? That is, just check whether this new block is at the end of the currently longest chain, and if it is then treat it as a "best" block. For the relay chain this should make no difference (because that's what the best block there is already), and for the parachains it would sidestep the issue you wrote about. (And since eventually we want to remove the concept of the best block from the client anyway I guess we'd have to do this anyway sooner or later?)

Does this sound good to you?

Reply:
Calculating what is the best block sounds like something that will break very fast :see_no_evil: 

As we want to remove the notion of what is the best block at some point anyway, why should we then only send for updates for best block(I didn't read through the entire old conversation we had in the initial issue). However, we currently send for all blocks and when we don't have this concept anymore, clients would have to change again to adapt to our changes?